### PR TITLE
Add AHRS to Tethys equipped model

### PR DESCRIPTION
**Pull request by hidmic**
Combining upstream IMU and magnetometer sensors. This is lacking a test but perhaps this is best tested as part of #151 e.g. make the AUV describe a known trajectory and sample IMU and magnetometer measurements along that trajectory to ensure values make sense.

### CODE REVIEW

**Comment by hidmic** [comment]
@braanan do we have a reference AHRS system from which to derive a noise model? Not that we have a lot of options -- we have gaussian or gaussian.

Also, I'd really need access to all of our repositories to complete AHRS integration.

**Comment by braanan** [comment]
I think we'll go with Gaussian then. Here's a link to the Sparton AHRS-M2 we have installed across the fleet: https://www.spartonnavex.com/product/ahrs-m2/ All the info you need should be in the datasheet.

As for access, all checks are cleared — just waiting for final approval from our MT.

**Comment by hidmic** [comment]
Alright, after much digging, I think I extracted all that can be extracted from the datasheet. There's no characterization of the magnetometer whatsoever, and IIUC I'm lacking data to use the bias drift model in Ignition.

File: lrauv_ignition_plugins/test/test_ahrs.cc
Comment (hidmic) [inline lrauv_ignition_plugins/test/test_ahrs.cc:216]: @braanan @arjo129 at first I tried to break this test into smaller ones, but I'm unable to instantiate `ignition::gazebo::TestFixture` more than once in the same process. It aborts with `CHECK failed: GeneratedDatabase()->Add(...)`, which appears to be coming from the shared library that bears `lrauv`
protobuf messages. I suspect the Ignition plugin system is attempting to reload it upon fixture construction (see [here](https://groups.google.com/g/protobuf/c/RZRCnOywdSk/m/s5QvifLCBgAJ?utm_medium=email&utm_source=footer)).

File: lrauv_ignition_plugins/test/test_ahrs.cc
Comment (hidmic) [inline lrauv_ignition_plugins/test/test_ahrs.cc:85]: @braanan @arjo129 I forced the model to be static to avoid motion whenever I put the AUV in unusual positions. Unfortunately, it also means we are not testing AHRS output when moving. It'd be tricky nonetheless. Hydrodynamics modelling gets in the way and it's hard to tell what the angular velocity and linear acceleration is going to be.

File: lrauv_ignition_plugins/test/test_ahrs.cc
Comment (arjo129) [reply lrauv_ignition_plugins/test/test_ahrs.cc:216]: Oh yeah I forgot to mention I think we keep each test isolated in a separate process as there was a bug in the way we are linking the messages. For whatever reason if we rely on external messages google protobuf will go :boom:. I have not been able to isolate exactly why this is the case.

File: lrauv_ignition_plugins/test/test_ahrs.cc
Comment (arjo129) [reply lrauv_ignition_plugins/test/test_ahrs.cc:85]: I think we agreed that as long as the AHRS is reporting the correct coordinate frame we should hope for the best with the gyro.

Comment (hidmic) [reply lrauv_ignition_plugins/test/test_ahrs.cc:85]: That's right. Thing is, I cannot test for gyro frame conventions if it's output is always zero. I guess we _are_ testing that the body frame is correct for the accelerometer so it should be good for the gyro too.

**Review by arjo129 — APPROVED** [review]
Hey @hidmic thanks for the tests and the imu. I think this is good to merge. For integration one possibility is to use the [`lrauv_state`](https://github.com/osrf/lrauv/blob/main/lrauv_ignition_plugins/proto/lrauv_state.proto) to pass the imu outputs. This structure is populated in [`TethysCommPlugin.cc`](https://github.com/osrf/lrauv/blob/main/lrauv_ignition_plugins/src/TethysCommPlugin.cc). In particular the `ratePQR` field may be of interest.

**Comment by arjo129** [comment]
@hidmic Should I merge this?

**Comment by hidmic** [comment]
> @hidmic Should I merge this?

@arjo129 need to make sure I can integrate it with the other half of the codebase first. Hopefully this week.

**Comment by hidmic** [comment]
@braanan @arjo129 this one's ready for a final review. Tests are passing locally. Integration with the rest of the code seems functional (though I see Tethys going for a run as #167 reports).

**Comment by hidmic** [comment]
Integration tests are failing upon installing `sdformat` :thinking:

**Comment by braanan** [comment]
@hidmic looks like there's a reverse sign in the accelerometer data - FSK should have positive down. 

I'm seeing:
```
2022-03-16T18:57:25.334Z,1647457045.334 [AHRS_M2](INFO): Accel:   X: 3.318425, Y: 0.055010, Z: -9.219960 (m/s2)
```

**Comment by hidmic** [comment]
@braanan my (perhaps vague) understanding of MEMS IMUs is that when standing still under constant gravity, its proof mass is displaced as if it was accelerating in the opposite direction (ie. upwards). This matches Ignition's IMU sensor behavior (see [here](https://github.com/ignitionrobotics/ign-sensors/blob/f2eb097977c35cd9ac6b81d5a5b51dd5bfeb94e7/src/ImuSensor.cc#L231) and gravity defaults [here](https://github.com/ignitionrobotics/sdformat/blob/c496c5bb787c25c6f3c2f5d10304da2a9efaafd6/src/World.cc#L57)) 

Having said that, I have no first-hand experience with Sparton's AHRS-M2 so perhaps it behaves differently.

**Comment by braanan** [comment]
It does. Here's the output from an actual vehicle
```
2022-03-16T19:08:03.896Z,1647457683.896 [AHRS_M2](INFO): PITCH: -0.384570, ROLL: -3.744142, YAW: 213.573527 (deg)
2022-03-16T19:08:03.897Z,1647457683.897 [AHRS_M2](INFO): Accel:   X: 0.064123, Y: -0.640986, Z: 9.850662 (m/s2)
```
I'm not sure if this output format is true to all AHRS models or unique to the M2. I'll check, since that will inform where we should implement the sign flip. If it's indeed true for all AHRS models we can see about reversing the sign in the plugin, otherwise, the sign flip should go in the AHRS_M2 device driver in the lrauv-application.

**Comment by braanan** [comment]
OK, I've asked around and it seems like there's not much consistency between vendors, so I'll add the sign flip in the AHRS_M2 driver with a comment.

**Comment by hidmic** [comment]
> OK, I've asked around and it seems like there's not much consistency between vendors, so I'll add the sign flip in the AHRS_M2 driver with a comment.

@braanan hmm, will that work? It's not that the linear acceleration Z-axis is flipped, only gravity is. I'll go ahead and guess that the real AHRS is estimating gravity's magnitude and direction and re-applying it to provide these acceleration measurements. Maybe we have to do the same in the plugin :/

**Comment by braanan** [comment]
@hidmic Looks like the imu plugin assumes that the linear accelarations (and the sensor) are alligned with the ENU frame of refrence. As far as I can tell, the rotation to NED is only applied to the oriantation data [here](https://github.com/ignitionrobotics/ign-sensors/blob/f2eb097977c35cd9ac6b81d5a5b51dd5bfeb94e7/src/ImuSensor.cc#L262), so we'll have to add that rotation to NED (FSK) to the linear accelarations and angular velocities in the plugin or take care of that on the LRAUV side. Maybe I'm missing something. 

Since this is an upstream plugin the second option will be simpler I think.

**Comment by hidmic** [comment]
@braanan may I get an approval?

**Review by braanan — APPROVED** [review]

**Pull request merged**
